compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1408,81 +1409,188 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                        ("a", "Result", "Err", ret_ty_matches(sym::Result))
-                    } else if Some(kind.did()) == tcx.get_diagnostic_item(sym::Option) {
-                        ("an", "Option", "None", ret_ty_matches(sym::Option))
-                    } else {
-                        return;
-                    };
-                if question {
-                    err.span_suggestion_verbose(
-                        expr.span.shrink_to_hi(),
-                        format!(
-                            "use the `?` operator to extract the `{self_ty}` value, propagating \
-                            {article} `{kind}::{variant}` value to the caller"
-                        ),
-                        "?",
-                        Applicability::MachineApplicable,
-                    );
+                let name = self.ty_to_value_string(actual);
+                let inner_id = kind.did();
+                let mutable = if let Some(AutorefOrPtrAdjustment::Autoref { mutbl, .. }) =
+                    pick.autoref_or_ptr_adjustment
+                {
+                    Some(mutbl)
                 } else {
+                    None
+                };
+
+                if tcx.is_diagnostic_item(sym::LocalKey, inner_id) {
+                    err.help("use `with` or `try_with` to access thread local storage");
+                } else if Some(kind.did()) == tcx.lang_items().maybe_uninit() {
+                    err.help(format!(
+                        "if this `{name}` has been initialized, \
+                        use one of the `assume_init` methods to access the inner value"
+                    ));
+                } else if tcx.is_diagnostic_item(sym::RefCell, inner_id) {
+                    match mutable {
+                        Some(Mutability::Not) => {
+                            err.span_suggestion_verbose(
+                                expr.span.shrink_to_hi(),
+                                format!(
+                                    "use `.borrow()` to borrow the `{ty}`, \
+                                    panicking if any outstanding mutable borrows exist."
+                                ),
+                                ".borrow()",
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                        Some(Mutability::Mut) => {
+                            err.span_suggestion_verbose(
+                                expr.span.shrink_to_hi(),
+                                format!(
+                                    "use `.borrow_mut()` to mutably borrow the `{ty}`, \
+                                    panicking if any outstanding borrows exist."
+                                ),
+                                ".borrow_mut()",
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
```

Comment:
Perhaps deduplicate a bit?

```suggestion
                        Some(mutbl) => {
                            let (suggestion, mutbly, mutbl) = match mutbl {
                                Mutability::Not => (".borrow()", "", ""),
                                Mutability::Mut => (".borrow()", "mutably ", "mutable"),
                            };
                            err.span_suggestion_verbose(
                                expr.span.shrink_to_hi(),
                                format!(
                                    "use `.borrow()` to {mutbly}borrow the `{ty}`, \
                                    panicking if any outstanding {mutbl}borrows exist."
                                ),
                                suggestion,
                                Applicability::MaybeIncorrect,
                            );
                        }
```

Reply:
Hmm I'm not sure. I prefer being too verbose rather than too clever, and this looks very clever  😄

Reply:
This is a very common pattern to deduplicate essentially identical diagnostic messages. I'd actually prefer if you deduplicated it even _more_, but this one is the easiest to apply.